Hi @daniwe4 
Sorry for the late response!
I just checked your PR (and rebased it locally fist): while running a `php setup/cli.php update -y` I get errors like ` [ERROR] Call to undefined method ilMailDatabaseUpdateSteps::step_2()`. the class `ilMailDatabaseUpdateSteps` currently only has one step and the `step_2`does not exist. could you please have a look at this? thanks a lot!

--- a/Modules/StudyProgramme/classes/Setup/ilStudyProgrammeUpdateAgent.php
+++ b/Modules/StudyProgramme/classes/Setup/ilStudyProgrammeUpdateAgent.php
@@ -1,6 +1,8 @@
 <?php declare(strict_types=1);
 
 use ILIAS\Setup;
+use ILIAS\Setup\Objective;
+use ILIAS\Setup\Metrics;
 
 class ilStudyProgrammeUpdateAgent extends Setup\Agent\NullAgent
 {
@@ -9,5 +11,10 @@ class ilStudyProgrammeUpdateAgent extends Setup\Agent\NullAgent
         return new ilDatabaseUpdateStepsExecutedObjective(
             new ilStudyProgrammeProgressTableUpdateSteps()
         );
+    }
+
+    public function getStatusObjective(Metrics\Storage $storage) : Objective
+    {
+        return new \ilDatabaseUpdateStepsMetricsCollectedObjective($storage, new ilStudyProgrammeProgressTableUpdateSteps());
     }
 }

--- a/Modules/TestQuestionPool/Setup/class.ilTestQuestionPoolSetupAgent.php
+++ b/Modules/TestQuestionPool/Setup/class.ilTestQuestionPoolSetupAgent.php
@@ -3,11 +3,17 @@
 use ILIAS\Setup\Agent\NullAgent;
 use ILIAS\Setup\Config;
 use ILIAS\Setup\Objective;
+use ILIAS\Setup\Metrics;
 
 class ilTestQuestionPoolSetupAgent extends NullAgent
 {
     public function getUpdateObjective(ILIAS\Setup\Config $config = null) : ILIAS\Setup\Objective
     {
         return new ilDatabaseUpdateStepsExecutedObjective(new ilTestQuestionPool80DBUpdateSteps());
+    }
+
+    public function getStatusObjective(Metrics\Storage $storage) : Objective
+    {
+        return new \ilDatabaseUpdateStepsMetricsCollectedObjective($storage, new ilTestQuestionPool80DBUpdateSteps());
     }
 }

--- a/Services/COPage/classes/Setup/class.Agent.php
+++ b/Services/COPage/classes/Setup/class.Agent.php
@@ -16,6 +16,8 @@
 namespace ILIAS\COPage\Setup;
 
 use ILIAS\Setup;
+use ILIAS\Setup\Objective;
+use ILIAS\Setup\Metrics;
 
 /**
  * @author Alexander Killing <killing@leifos.de>
@@ -25,5 +27,10 @@ class Agent extends Setup\Agent\NullAgent
     public function getUpdateObjective(Setup\Config $config = null) : Setup\Objective
     {
         return new \ilDatabaseUpdateStepsExecutedObjective(new ilCOPageDBUpdateSteps());
+    }
+
+    public function getStatusObjective(Metrics\Storage $storage) : Objective
+    {
+        return new \ilDatabaseUpdateStepsMetricsCollectedObjective($storage, new ilCOPageDBUpdateSteps());
     }
 }

--- a/Services/Database/classes/Setup/class.ilDBStepReader.php
+++ b/Services/Database/classes/Setup/class.ilDBStepReader.php
@@ -1,0 +1,39 @@
+<?php declare(strict_types=1);
+
+/* Copyright (c) 2021 - Daniel Weise <daniel.weise@concepts-and-training.de> - Extended GPL, see LICENSE */
+
+class ilDBStepReader
+{
+    /**
+     * Get the number of the latest database step in this class.
+     */
+    public function getLatestStepNumber(string $step_class_name, string $step_prefix) : int
+    {
+        $steps = $this->readStepNumbers($step_class_name, $step_prefix);
+        return $steps[count($steps) - 1];
+    }
+
+    /**
+     * Get a list of all steps in this class.
+     */
+    public function readStepNumbers(string $step_class_name, string $step_prefix) : array
+    {
+        $step_numbers = [];
+        foreach (get_class_methods($step_class_name) as $method) {
+            if (stripos($method, $step_prefix) !== 0) {
+                continue;
+            }
+
+            $number = substr($method, strlen($step_prefix));
+
+            if (!preg_match("/^[1-9]\d*$/", $number)) {
+                throw new LogicException("Method $method seems to be a step but has an odd looking number");
+            }
+
+            $step_numbers[] = (int) $number;
+        }
+
+        sort($step_numbers);
+        return $step_numbers;
+    }
+}

--- a/Services/Database/classes/Setup/class.ilDBStepReaderExistsObjective.php
+++ b/Services/Database/classes/Setup/class.ilDBStepReaderExistsObjective.php
@@ -1,0 +1,48 @@
+<?php
+
+/* Copyright (c) 2021 - Daniel Weise <daniel.weise@concepts-and-training.de> - Extended GPL, see LICENSE */
+
+use ILIAS\Setup;
+
+class ilDBStepReaderExistsObjective implements Setup\Objective
+{
+    public function getHash() : string
+    {
+        return hash("sha256", self::class);
+    }
+
+    public function getLabel() : string
+    {
+        return "The step reader for database update steps exists.";
+    }
+
+    public function isNotable() : bool
+    {
+        return true;
+    }
+
+    public function getPreconditions(Setup\Environment $environment) : array
+    {
+        return [
+            new ilDatabaseUpdatedObjective()
+        ];
+    }
+
+    public function achieve(Setup\Environment $environment) : Setup\Environment
+    {
+        return $environment
+            ->withResource(
+                ilDBStepReader::class,
+                new ilDBStepReader()
+            );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function isApplicable(Setup\Environment $environment) : bool
+    {
+        $execution_db = $environment->getResource(ilDBStepReader::class);
+        return is_null($execution_db);
+    }
+}

--- a/Services/Database/classes/Setup/class.ilDatabaseUpdateStepsExecutedObjective.php
+++ b/Services/Database/classes/Setup/class.ilDatabaseUpdateStepsExecutedObjective.php
@@ -15,11 +15,9 @@ class ilDatabaseUpdateStepsExecutedObjective implements Objective
 
     protected ilDatabaseUpdateSteps $steps;
     protected string $steps_class;
-    protected ?array $step_numbers = null;
 
-    public function __construct(
-        ilDatabaseUpdateSteps $steps
-    ) {
+    public function __construct(ilDatabaseUpdateSteps $steps)
+    {
         $this->steps = $steps;
         $this->steps_class = get_class($this->steps);
     }
@@ -32,7 +30,7 @@ class ilDatabaseUpdateStepsExecutedObjective implements Objective
     {
         return hash(
             "sha256",
-            $this->steps_class
+            self::class . $this->steps_class
         );
     }
 
@@ -55,8 +53,9 @@ class ilDatabaseUpdateStepsExecutedObjective implements Objective
     public function getPreconditions(Environment $environment) : array
     {
         return [
-            new \ilDBStepExecutionDBExistsObjective(),
-            new \ilDatabaseUpdatedObjective()
+            new ilDBStepExecutionDBExistsObjective(),
+            new ilDatabaseUpdatedObjective(),
+            new ilDBStepReaderExistsObjective()
         ];
     }
 
@@ -65,7 +64,8 @@ class ilDatabaseUpdateStepsExecutedObjective implements Objective
      */
     public function achieve(Environment $environment) : Environment
     {
-        $execution_log = $environment->getResource(\ilDatabaseUpdateStepExecutionLog::class);
+        $execution_log = $environment->getResource(ilDatabaseUpdateStepExecutionLog::class);
+        $step_reader = $environment->getResource(ilDBStepReader::class);
 
         $last_started_step = $execution_log->getLastStartedStep($this->steps_class);
         $last_finished_step = $execution_log->getLastFinishedStep($this->steps_class);
@@ -76,15 +76,15 @@ class ilDatabaseUpdateStepsExecutedObjective implements Objective
             );
         }
 
-        if ($last_finished_step === $this->getLatestStepNumber()) {
+        if ($last_finished_step === $step_reader->getLatestStepNumber($this->steps_class, self::STEP_METHOD_PREFIX)) {
             return $environment;
         }
 
         $db = $environment->getResource(Environment::RESOURCE_DATABASE);
         $this->steps->prepare($db);
 
-        $this->readSteps();
-        foreach ($this->step_numbers as $step) {
+        $steps = $step_reader->readStepNumbers($this->steps_class, self::STEP_METHOD_PREFIX);
+        foreach ($steps as $step) {
             if ($step <= $last_finished_step) {
                 continue;
             }
@@ -102,55 +102,21 @@ class ilDatabaseUpdateStepsExecutedObjective implements Objective
      */
     public function isApplicable(Environment $environment) : bool
     {
-        $execution_log = $environment->getResource(\ilDatabaseUpdateStepExecutionLog::class);
-        return $execution_log->getLastFinishedStep($this->steps_class) !== $this->getLatestStepNumber();
+        $execution_log = $environment->getResource(ilDatabaseUpdateStepExecutionLog::class);
+        $step_reader = $environment->getResource(ilDBStepReader::class);
+
+        return $execution_log->getLastFinishedStep($this->steps_class) !== $step_reader->getLatestStepNumber(
+            $this->steps_class,
+            self::STEP_METHOD_PREFIX
+        );
     }
 
     protected function throwStepNotFinishedException(int $started, int $finished) : void
     {
-        throw new \RuntimeException(
+        throw new RuntimeException(
             "For update steps in $this->steps_class: step $started was started " .
             "last, but step $finished was finished last. Aborting because of that " .
             "mismatch."
         );
-    }
-
-    /**
-     * Get the number of latest database step in this class.
-     */
-    final public function getLatestStepNumber() : int
-    {
-        $this->readSteps();
-        return $this->step_numbers[count($this->step_numbers) - 1];
-    }
-
-    /**
-     * Get a list of all steps in this class.
-     *
-     * @return int[]
-     */
-    protected function readSteps() : void
-    {
-        if (!is_null($this->step_numbers)) {
-            return;
-        }
-
-        $this->step_numbers = [];
-
-        foreach (get_class_methods($this->steps_class) as $method) {
-            if (stripos($method, self::STEP_METHOD_PREFIX) !== 0) {
-                continue;
-            }
-
-            $number = substr($method, strlen(self::STEP_METHOD_PREFIX));
-
-            if (!preg_match("/^[1-9]\d*$/", $number)) {
-                throw new \LogicException("Method $method seems to be a step but has an odd looking number");
-            }
-
-            $this->step_numbers[] = (int) $number;
-        }
-
-        sort($this->step_numbers);
     }
 }

--- a/Services/Database/classes/Setup/class.ilDatabaseUpdateStepsMetricsCollectedObjective.php
+++ b/Services/Database/classes/Setup/class.ilDatabaseUpdateStepsMetricsCollectedObjective.php
@@ -1,0 +1,74 @@
+<?php declare(strict_types=1);
+
+/* Copyright (c) 2021 - Daniel Weise <daniel.weise@concepts-and-training.de> - Extended GPL, see LICENSE */
+
+use ILIAS\Setup;
+use ILIAS\Setup\Metrics\Metric;
+use ILIAS\Setup\Metrics\Storage;
+
+class ilDatabaseUpdateStepsMetricsCollectedObjective extends Setup\Metrics\CollectedObjective
+{
+    const STEP_METHOD_PREFIX = "step_";
+
+    protected string $step_class;
+
+    public function __construct(Storage $storage, ilDatabaseUpdateSteps $steps)
+    {
+        parent::__construct($storage);
+        $this->step_class = get_class($steps);
+    }
+
+    protected function collectFrom(Setup\Environment $environment, Storage $storage) : void
+    {
+        $execution_log = $environment->getResource(ilDatabaseUpdateStepExecutionLog::class);
+        $step_reader = $environment->getResource(ilDBStepReader::class);
+
+        $version = new Metric(
+            Metric::STABILITY_STABLE,
+            Metric::TYPE_TEXT,
+            (string) ($execution_log->getLastFinishedStep($this->step_class))
+        );
+
+        $available_version = new Metric(
+            Metric::STABILITY_STABLE,
+            Metric::TYPE_TEXT,
+            (string) $step_reader->getLatestStepNumber($this->step_class, self::STEP_METHOD_PREFIX)
+        );
+
+        $update_required = new Metric(
+            Metric::STABILITY_STABLE,
+            Metric::TYPE_BOOL,
+            $execution_log->getLastFinishedStep($this->step_class) !== $step_reader->getLatestStepNumber(
+                $this->step_class,
+                self::STEP_METHOD_PREFIX
+            )
+        );
+
+        $collection = new Metric(
+            Metric::STABILITY_STABLE,
+            Metric::TYPE_COLLECTION,
+            [
+                "version" => $version,
+                "available_version" => $available_version,
+                "update_required" => $update_required
+            ]
+        );
+
+        $storage->store($this->step_class, $collection);
+    }
+
+    protected function getTentativePreconditions(Setup\Environment $environment) : array
+    {
+        return [
+            new ilIniFilesLoadedObjective(),
+            new ilDatabaseInitializedObjective(),
+            new ilDBStepExecutionDBExistsObjective(),
+            new ilDBStepReaderExistsObjective()
+        ];
+    }
+
+    public function getHash() : string
+    {
+        return hash("sha256", static::class . $this->step_class);
+    }
+}

--- a/Services/Database/test/Setup/ilDBStepReaderTest.php
+++ b/Services/Database/test/Setup/ilDBStepReaderTest.php
@@ -1,0 +1,65 @@
+<?php declare(strict_types=1);
+
+/* Copyright (c) 2021 - Daniel Weise <daniel.weise@concepts-and-training.de> - Extended GPL, see LICENSE */
+
+use PHPUnit\Framework\TestCase;
+
+class ilDBStepReaderTestObject extends ilDBStepReader
+{
+    public function setStepNumbers(array $arr) : void
+    {
+        $this->step_numbers = $arr;
+    }
+}
+
+class Test_ilDBStepReader implements ilDatabaseUpdateSteps
+{
+    public $called = [];
+
+    protected ?ilDBInterface $db = null;
+
+    public function prepare(ilDBInterface $db)
+    {
+        $this->db = $db;
+    }
+
+
+    public function step_1()
+    {
+    }
+
+    // 4 comes before 2 to check if the class gets the sorting right
+    public function step_4()
+    {
+    }
+
+    public function step_2()
+    {
+    }
+}
+
+class ilDBStepReaderTest extends TestCase
+{
+    public function testObjectCreation() : void
+    {
+        $obj = new ilDBStepReader();
+        $this->assertInstanceOf(ilDBStepReader::class, $obj);
+    }
+
+    public function test_getLatestStepNumber() : void
+    {
+        $obj = new ilDBStepReaderTestObject();
+        $this->assertEquals(4, $obj->getLatestStepNumber(Test_ilDBStepReader::class, "step_"));
+    }
+
+    public function test_readSteps() : void
+    {
+        $obj = new ilDBStepReaderTestObject();
+        $result = $obj->readStepNumbers(Test_ilDBStepReader::class, "step_");
+
+        $this->assertIsArray($result);
+        $this->assertEquals(1, $result[0]);
+        $this->assertEquals(2, $result[1]);
+        $this->assertEquals(4, $result[2]);
+    }
+}

--- a/Services/Database/test/Setup/ilDatabaseSetupSuite.php
+++ b/Services/Database/test/Setup/ilDatabaseSetupSuite.php
@@ -6,15 +6,21 @@ use PHPUnit\Framework\TestSuite;
 
 class ilDatabaseSetupSuite extends TestSuite
 {
-    public static function suite() : \ilDatabaseSetupSuite
+    public static function suite() : ilDatabaseSetupSuite
     {
         $suite = new self();
 
         require_once(__DIR__ . "/ilDatabaseUpdateStepsExecutedObjectiveTest.php");
-        $suite->addTestSuite(\ilDatabaseUpdateStepsExecutedObjectiveTest::class);
+        $suite->addTestSuite(ilDatabaseUpdateStepsExecutedObjectiveTest::class);
+
+        require_once(__DIR__ . "/ilDBStepReaderTest.php");
+        $suite->addTestSuite(ilDBStepReaderTest::class);
 
         require_once(__DIR__ . "/ilDBStepExecutionDBTest.php");
-        $suite->addTestSuite(\ilDBStepExecutionDBTest::class);
+        $suite->addTestSuite(ilDBStepExecutionDBTest::class);
+
+        require_once __DIR__ . "/ilDatabaseUpdateStepsMetricsCollectedObjectiveTest.php";
+        $suite->addTestSuite(ilDatabaseUpdateStepsMetricsCollectedObjectiveTest::class);
 
         return $suite;
     }

--- a/Services/Database/test/Setup/ilDatabaseUpdateStepsExecutedObjectiveTest.php
+++ b/Services/Database/test/Setup/ilDatabaseUpdateStepsExecutedObjectiveTest.php
@@ -6,16 +6,14 @@ use PHPUnit\Framework\TestCase;
 
 use ILIAS\Setup\Environment;
 use ILIAS\Setup\ArrayEnvironment;
-use ILIAS\Setup\ObjectiveIterator;
-use ILIAS\Setup\Objective;
 
 class Test_ilDatabaseUpdateSteps implements ilDatabaseUpdateSteps
 {
     public $called = [];
 
-    protected ?\ilDBInterface $db = null;
+    protected ?ilDBInterface $db = null;
 
-    public function prepare(\ilDBInterface $db)
+    public function prepare(ilDBInterface $db)
     {
         $this->db = $db;
     }
@@ -76,8 +74,11 @@ class ilDatabaseUpdateStepsExecutedObjectiveTest extends TestCase
                 return 0;
             }
         };
-        $db = $this->createMock(\ilDBInterface::class);
+        $steps_reader = new class() extends ilDBStepReader {
+        };
+        $db = $this->createMock(ilDBInterface::class);
         $env = new ArrayEnvironment([
+            ilDBStepReader::class => $steps_reader,
             ilDatabaseUpdateStepExecutionLog::class => $execution_log,
             Environment::RESOURCE_DATABASE => $db
         ]);
@@ -116,8 +117,11 @@ class ilDatabaseUpdateStepsExecutedObjectiveTest extends TestCase
                 return 0;
             }
         };
-        $db = $this->createMock(\ilDBInterface::class);
+        $steps_reader = new class() extends ilDBStepReader {
+        };
+        $db = $this->createMock(ilDBInterface::class);
         $env = new ArrayEnvironment([
+            ilDBStepReader::class => $steps_reader,
             ilDatabaseUpdateStepExecutionLog::class => $execution_log,
             Environment::RESOURCE_DATABASE => $db
         ]);
@@ -157,8 +161,11 @@ class ilDatabaseUpdateStepsExecutedObjectiveTest extends TestCase
                 return 1;
             }
         };
-        $db = $this->createMock(\ilDBInterface::class);
+        $steps_reader = new class() extends ilDBStepReader {
+        };
+        $db = $this->createMock(ilDBInterface::class);
         $env = new ArrayEnvironment([
+            ilDBStepReader::class => $steps_reader,
             ilDatabaseUpdateStepExecutionLog::class => $execution_log,
             Environment::RESOURCE_DATABASE => $db
         ]);
@@ -173,7 +180,7 @@ class ilDatabaseUpdateStepsExecutedObjectiveTest extends TestCase
 
     public function testExceptionOnNonMatchingStartAndFinished()
     {
-        $this->expectException(\RuntimeException::class);
+        $this->expectException(RuntimeException::class);
 
         $execution_log = new class() implements ilDatabaseUpdateStepExecutionLog {
             public function started(string $class, int $step) : void
@@ -191,7 +198,7 @@ class ilDatabaseUpdateStepsExecutedObjectiveTest extends TestCase
                 return 1;
             }
         };
-        $db = $this->createMock(\ilDBInterface::class);
+        $db = $this->createMock(ilDBInterface::class);
         $env = new ArrayEnvironment([
             ilDatabaseUpdateStepExecutionLog::class => $execution_log,
             Environment::RESOURCE_DATABASE => $db

--- a/Services/Database/test/Setup/ilDatabaseUpdateStepsMetricsCollectedObjectiveTest.php
+++ b/Services/Database/test/Setup/ilDatabaseUpdateStepsMetricsCollectedObjectiveTest.php
@@ -1,0 +1,48 @@
+<?php declare(strict_types=1);
+
+/* Copyright (c) 2021 - Daniel Weise <daniel.weise@concepts-and-training.de> - Extended GPL, see LICENSE */
+
+use PHPUnit\Framework\TestCase;
+use ILIAS\Setup\Metrics\Storage;
+
+class Test_ilDatabaseUpdateSteps2 implements ilDatabaseUpdateSteps
+{
+    protected ?ilDBInterface $db = null;
+
+    public function prepare(ilDBInterface $db)
+    {
+        $this->db = $db;
+    }
+
+    public function step_1(ilDBInterface $db)
+    {
+    }
+
+    public function step_4(ilDBInterface $db)
+    {
+    }
+
+    public function step_2(ilDBInterface $db)
+    {
+    }
+}
+
+class ilDatabaseUpdateStepsMetricsCollectedObjectiveTest extends TestCase
+{
+    public Test_ilDatabaseUpdateSteps2 $steps;
+    public Storage $storage;
+
+    protected function setUp() : void
+    {
+        $this->steps = new Test_ilDatabaseUpdateSteps2;
+        $this->storage = $this->createMock(Storage::class);
+    }
+
+    public function testObjectCreation() : ilDatabaseUpdateStepsMetricsCollectedObjective
+    {
+        $obj = new ilDatabaseUpdateStepsMetricsCollectedObjective($this->storage, $this->steps);
+        $this->assertInstanceOf(ilDatabaseUpdateStepsMetricsCollectedObjective::class, $obj);
+
+        return $obj;
+    }
+}

--- a/docs/development/database-updates.md
+++ b/docs/development/database-updates.md
@@ -69,7 +69,22 @@ class MySetupAgent extends Setup\Agent\NullAgent
 }
 ``` 
 
-In the ilMyDBUpdateSteps you can add your consecutive steps by adding methods according
+To ensure that the setup/status command will output the current database step status
+of your component add the method `getStatusObjective` to your Agent. 
+
+```php
+use ILIAS\Setup;
+
+class MySetupAgent extends Setup\Agent\NullAgent
+{
+    public function getStatusObjective(Setup\Metrics\Storage $storage) : Setup\Objective
+    {
+        return new ilDatabaseUpdateStepsMetricsCollectedObjective($storage, new MyDBUpdateSteps());
+    }
+}
+``` 
+
+In the MyDBUpdateSteps you can add your consecutive steps by adding methods according
 to this schema:
 
 ```php

